### PR TITLE
Fixed Issue #18370: Password suggestions sometimes doens't match rules

### PR DESCRIPTION
**Pull request by gabrieljenik**
Fixed issue
Added a few methods on test base class for enhancing the testing of plugins.
Added tests for this issue (generate random) as well as for checkValidity (issue 18542)

### CODE REVIEW

File: application/libraries/PluginManager/PluginEvent.php [hunk 1]
```diff
@@ -173,6 +173,20 @@ public function set($key, $value)
         return $this;
     }
 
+    /**
+     * Set mass key/value pair to be used by plugins hanlding this event.
+     *
+     * @param array $values
+     * @return \PluginEvent Fluent interface
+     */
+    public function setFromArray($values)
```
Comment (Shnoulle) [inline application/libraries/PluginManager/PluginEvent.php:182]: Only used for test ?

Comment (gabrieljenik) [reply application/libraries/PluginManager/PluginEvent.php:182]: Right now, yes. New cases shall appear as it is available.

Comment (Shnoulle) [reply application/libraries/PluginManager/PluginEvent.php:182]: We already have API issue with only a single function, but you add more ?

Have such function disable all core fix see https://github.com/LimeSurvey/LimeSurvey/pull/2753 or https://github.com/LimeSurvey/LimeSurvey/pull/2743 …

Comment (Shnoulle) [reply application/libraries/PluginManager/PluginEvent.php:182]: I don't say there are conflict , clearly not **currently**. But in some month/year : there are potentially more API broke.

Why not, but i don't really find why adding this function is interesting. 

Maybe i'm wrong ?

Comment (gabrieljenik) [reply application/libraries/PluginManager/PluginEvent.php]: Don't agree. Still, already rolledback :)

File: application/helpers/common_helper.php [hunk 2]
```diff
@@ -5041,3 +5041,27 @@ function standardDeviation(array $numbers): float
...
     return sqrt($variance / $numberOfElements);
 }
+
+/**
+ * Returns a random number
+ * @param int $max The number maximun of string
+ * @param int $min The number minimum of string
+ * @return int
+ */
+function safeRandom($max, $min = 0)
```
Comment (Shnoulle) [inline application/helpers/common_helper.php:5051]: This 2 function are only used for this plugin, right ?

Adding it as private static function inside plugin seems better in my opinion.

**Review by Shnoulle — CHANGES_REQUESTED** [review]
I think it's better to have function for plugin inside plugin only.

Because usage of limesurvey global function are NOT in API ([common_helper](https://github.com/LimeSurvey/LimeSurvey/blob/master/application/helpers/common_helper.php), [frontend_helper](https://github.com/LimeSurvey/LimeSurvey/blob/master/application/helpers/frontend_helper.php) etc …)

We already broke global function without api update.

If you really think it must be in API : add it in API : https://github.com/LimeSurvey/LimeSurvey/blob/master/application/libraries/PluginManager/LimesurveyApi.php

My opinion : leave it in plugin only.

**Comment by gabrieljenik** [comment]
> I think it's better to have function for plugin inside plugin only.

I placed them in commons because they can be helpfull for others.
At the very beggining new functions will always be used on limited placed.
Still, thinking forward they can be usefull.

If all is private, nothing is reused, difficult to set standards, reuse code, and not reinvent the wheel.

> Adding it as private static function inside plugin seems better in my opinion.

I think code reviews are for reviewing stuff lines up with a global way of coding.
Is not for approving line by line.
 I understand you don't agree, but I believe we can live with some disagreements.

**Comment by gabrieljenik** [comment]
> We already broke global function without api update.

What does this mean?

File: application/helpers/common_helper.php (hunk 2 above)
Comment (olleharstedt) [reply application/helpers/common_helper.php:5051]: I don't care that much about private vs common but why do we need safe random at all? Can't we just use mt_rand? Or PHP version issue?

File: application/helpers/common_helper.php [hunk 3]
```diff
@@ -5041,3 +5041,27 @@ function standardDeviation(array $numbers): float
...
+{
+    if (function_exists('random_int')) {
+        return random_int($min, $max);
+    }
+    return mt_rand($min, $max);
+}
+
+/**
+ * Returns the character of a string
```
Comment (olleharstedt) [inline application/helpers/common_helper.php:5060]: "the" --> "a random"?

File: tests/TestBaseClass.php [hunk 4]
```diff
@@ -238,4 +256,22 @@ protected static function addUserPermissions(int $userId, array $permissions)
...
+     * Expected to be used for unit testing purposes on edge cases.
+     *
+     * @param object $object
+     * @param string $methodName
+     * @param array $parameters
+     * @return mixed
+     */
+    public function invokePrivateMethod(&$object, $methodName, array $parameters = array())
+    {
+        $reflection = new \ReflectionClass(get_class($object));
```
Comment (olleharstedt) [inline tests/TestBaseClass.php:271]: Ah this is pretty nice. We've started to use another mocking framework too to get around PHPUnit limitations.

File: application/helpers/common_helper.php [hunk 5]
```diff
@@ -5041,3 +5041,27 @@ function standardDeviation(array $numbers): float
 
     return sqrt($variance / $numberOfElements);
 }
+
+/**
+ * Returns a random number
+ * @param int $max The number maximun of string
```
Comment (olleharstedt) [inline application/helpers/common_helper.php:5047]: "string" is probably wrong here

**Comment by Shnoulle** [comment]
> > We already broke global function without api update.
> 
> What does this mean?

Updating function in common_helper didn't update API, then : if plugin used a common_helper function : maybe it can break during a minor update.

And it's already happen

**Comment by Shnoulle** [comment]
> > I think it's better to have function for plugin inside plugin only.
> 
> I placed them in commons because they can be helpfull for others. At the very beggining new functions will always be used on limited placed. Still, thinking forward they can be usefull.
> 
> If all is private, nothing is reused, difficult to set standards, reuse code, and not reinvent the wheel.

I know that : but it's here plugin, it must work during little fix, and use a function not in API is not something to do.


> > Adding it as private static function inside plugin seems better in my opinion.
> 
> I think code reviews are for reviewing stuff lines up with a global way of coding. Is not for approving line by line. I understand you don't agree, but I believe we can live with some disagreements.

But ? Did you think the same when looking my Pull Request ?

**Comment by gabrieljenik** [comment]
> But ? Did you think the same when looking my Pull Request ?

Yes, usually I try to.

I think is not common for me to ask for change requests.
I try to make them as an exceptional thing.

> If you think it can be interesting for other plugin : add it in API

I think it can be interesting not only for API, but for other stuff as well.

**Comment by Shnoulle** [comment]
> I think it can be interesting not only for API, but for other stuff as well.

Why not in global + in API.

But still : i close conversation. I write clearly my opinion about usage of core function out of API.

**Review by Shnoulle — APPROVED** [review]
Still think usage of core function must be avoided by API.

Except : no issue in code.

**Comment by gabrieljenik** [comment]
I am moving the code as static method. Issue solved.

**Comment by Shnoulle** [comment]
I already approved before

**Review by Shnoulle — APPROVED** [review]

File: tests/TestBaseClass.php (hunk 4 above)
Comment (Shnoulle) [reply tests/TestBaseClass.php:271]: Just a confirmation : this can not be used via **URL** hacking to call a private method and create security issue ?

**Comment by gabrieljenik** [comment]
> I already approved before

Yes, seems we didi it at almostthe same time

File: tests/TestBaseClass.php (hunk 4 above)
Comment (gabrieljenik) [reply tests/TestBaseClass.php:271]: I can change it to protected probably.

Comment (Shnoulle) [reply tests/TestBaseClass.php:271]: I really think it can not happen, only controller can be call directly. But why not protected, think it must be used only in test.

If needed elsewhere : move the function to public is better.

:+1:

File: application/helpers/common_helper.php (hunk 3 above)
Comment (gabrieljenik) [reply application/helpers/common_helper.php]: Updated

File: tests/TestBaseClass.php [hunk 6]
```diff
@@ -256,22 +256,4 @@ protected static function addUserPermissions(int $userId, array $permissions)
...
-    /**
-     * Invokes a private method from an object
-     * Expected to be used for unit testing purposes on edge cases.
-     *
-     * @param object $object
-     * @param string $methodName
-     * @param array $parameters
-     * @return mixed
-     */
-    public function invokePrivateMethod(&$object, $methodName, array $parameters = array())
```
Comment (Shnoulle) [inline tests/TestBaseClass.php]: Removed ?

File: tests/TestBaseClass.php (hunk 4 above)
Comment (gabrieljenik) [reply tests/TestBaseClass.php]: At last, I removed it. I think I initially used it and then just used the events.
When needed can be restablished

File: tests/TestBaseClass.php (hunk 6 above)
Comment (gabrieljenik) [reply tests/TestBaseClass.php]: https://github.com/LimeSurvey/LimeSurvey/pull/2848#discussion_r1073555751

File: application/helpers/common_helper.php (hunk 5 above)
Comment (gabrieljenik) [reply application/helpers/common_helper.php]: Updated

**Comment by olleharstedt** [comment]
Password unit test fails, it seems?

**Comment by gabrieljenik** [comment]
> Password unit test fails, it seems?

he he...
Thats because this test relies on a fix pushed on this PR:
https://github.com/LimeSurvey/LimeSurvey/pull/2847#top

Thing is I discovered the issue #18542 while doing the password auto tests.
It seemed to me it was OK to do it as a separate issue.

**Comment by olleharstedt** [comment]
Still fails?

**Comment by gabrieljenik** [comment]
> Still fails?

It was an old run. Need to re run it.

**Comment by gabrieljenik** [comment]
> tter to have function for plugin inside pl

Now Pull is green :)

**Comment by olleharstedt** [comment]
> > tter to have function for plugin inside pl
> 
> Now Pull is green :)

You sure?

**Comment by gabrieljenik** [comment]
Thats because this test relies on a fix pushed on this PR:
https://github.com/LimeSurvey/LimeSurvey/pull/2847

Push is failing. Pull is good. Should I merge master as to have all green?

**Comment by gabrieljenik** [comment]
Now is green green
